### ISSUE TITLE
Fermat Enums code upgrade/enhacement.

### ISSUE THREAD

**Issue by Yayotron** [Fermat-ORG/fermat#2575]
In the current standard, any enum that inherits from FermatEnum has a method FermatEnum#getByCode(String) which looks something like this:
```
public static FermatEnum getByCode(String code) throws InvalidParameterException {
   switch (code){
     case "code1"
          return FermatEnum.enum1;
     case "code2":
          return FermatEnum.enum2;
     case "n":
          return FermatEnum.enumn;
     default:
          throw new InvalidParameterException(InvalidParameterException.DEFAULT_MESSAGE, null, "Code Received: " + code, "This Code Is Not Valid for the EventStatus enum.");
        }
    }
```
That code is hard to maintain: if you add a new enum you'll need to add it on that method too, also if you forgot to add the case on the switch statement the exception will be thrown. 
That's against single responsibility principle, one of the main OOP principles.

Suggested solution:

```
for (EventStatus status : values()) {
    if (status.getCode().equals(code)) return status;
}
throw new InvalidParameterException(InvalidParameterException.DEFAULT_MESSAGE, null, "Code Received: " + code, "This Code Is Not Valid for the EventStatus enum.");

/**
* values() is a method that every enum has, that method returns an array with all
*  the different instances for that enum type.
*/
```

@acostarodrigo label this and tag the people for discuss this, please.

**Comment by lnacosta** [comment]
Please, we should end this debate, answer me. @Yayotron @acostarodrigo

**Comment by Yayotron** [comment]
Well, a repeated code on a enum is something that rarely happen, just like changing a code and forgot about changing the switch, or vice versa, but both are the exception not the rule. 
Also, the performance improvement for the for statement is tiny, nothing that would change fermat forever; the most significant enhacement is a code maintenance and readability.
So the proposed change will in most part affect us and not the final user, it's your decision if we should implement it, I still vote for the yes.

**Comment by acostarodrigo** [comment]
I also agree.
A change in performance for this change I don't think will have any impact.
And I had cases to solve in which the code in the enum was not correct, so I would prefer to have it automated.

**Comment by Yayotron** [comment]
Hey, it have been a while and I want to close this issue.
Can we end up this discussion? I still hold my position, and have been working with my suggested solution for a while. Is there really a reason to keep the switch? We already have applied lots of solutions for the compile-time problem.

**Comment by furszy** [comment]
@Yayotron with large enums we cant do this, every class that you add in an enum object is a new class in memory and one more iterate that we have to do until found the value that we are looking for in a getFrom method.
Close this issue please.

**Comment by lnacosta** [comment]
@furszy I didn't understand, what's the difference between doing a for loop, and ask the same number of times if it equals?

**Comment by acostarodrigo** [comment]
I also think we should implement @Yayotron solution.
I spend an entire day investigating a new plugin issue and the cause was that I forgot to add the enum in the return statement.

**Comment by furszy** [comment]
@lnacosta performance reasons, this is why the switch statement exist :P, read any documentation about it, the while statement have a thread blocker and so on.
We have to look for another approach to solve this problem. Maybe the solution is not using an enum class and register plugins at runtine with a previos verification.

**Comment by Yayotron** [comment]
I think that's micro optimization and premature optimization. Thing that should be avoided, even when it is true about the thread lock, and in theory is _faster_ in the practice that difference is humanly impossible to detect, it takes about nano seconds to process a for cycle with millions of elements. 
We can benchmark it, yet the difference won't be observable. The main difference is about code readability/avoid bugs, not performance, as I stated in the past.

**Comment by furszy** [comment]
We have to start doing optimizations in every opportunity that we have. I know that is a small one now but this kind of thing will be replicated in every enum in Fermat and we can not agree with that.
I promise that i will improve it when i finish my todo list.

**Comment by acostarodrigo** [comment]
I agree that the improvement will be in code readability. So I think we need to change it-

What is your proposal regarding Enums? @Yayotron may be able to do it now for you.

**Comment by Yayotron** [comment]
You should write it in there, that's why this issue is for. Discuss and improve our enum pattern.

**Comment by furszy** [comment]
We have to have the micro licences active or any auth structure created that is why i am saying that is for the future...
Que tenga mejor lectura lo entiendo Rodri pero eso no quiere decir que este bien, todo lo que podamos mejorar/dejar bien hay que hacerlo por más mínimo que sea para no se extienda por Fermat, ya nos pasó con los ns y eso fue un desastre. No es una pelea por ver quien la tiene más grande esto, es para que no se copien cosas mal después de mi parte.

**Comment by furszy** [comment]
Sorry but i have to say it in spanish to say what i want with the real words that i was looking for.

**Comment by Yayotron** [comment]
Being 3 nanoseconds faster neither makes it to be good. 
It is all about the pros and cons, the switch statement only provides a slightly performance improvement (and as I said, we should benchmark this, the performance improvement at this scale, even while being used by thousands of enums at once won't change anything) while losing a lot of code readability, increasing the maintenance cost a lot, and opening a huge space for bugs, which are hard to track. As you said, it's not about winning the discussion and having the reason. It is about improving the current format, not saying my solution is the right one, but stating that it is way better than the current one.

**Comment by acostarodrigo** [comment]
Estas siendo cabeza dura me parece...Leon implementó el ```for``` con el ```values```  en el enum de ```Plugins``` y había como 5 plugines de CBP que tenian mal puesto el enum y eso se detectó por el cambio.

Debe haber menos de 10 Enums grandes donde hace falta cambiarlo para evitar errores y pérdida de tiempo de la gente debuggeando. La diferencia de performance debe ser imposible de medir al cambiar esto.

**Comment by furszy** [comment]
No quiero que se repita por todos lados y que después querer optimizarlo sea un gasto increíble de tiempo, es verdad que no medimos cuanto es el consumo pero viendo como se acostumbramos (todos t me incluyo) a repetir las cosas acá, va a ser extremadamente difícil de cambiar si en algún momento queremos hacerlo.

**Comment by furszy** [comment]
Lo digo porque por ejemplo ahora tengo que cambiar los agentes de Fermat y es terrible, todos usamos el mismo tipo y lo armamos de la misma forma.

**Comment by Yayotron** [comment]
Ése siempre es el resultado de un mal diseño inicial, retrabajo en un futuro cuando se quiera mejorar. Por eso estamos acá discutiendo cuál es el mejor patrón (sí, para que todos se lo copien, porque ya se discutió y se supone que es la mejor manera de hacerlo) para nuestros enums. Capaz la solución que tenés en mente tampoco es la mejor, o la que yo propuse y terminamos aplicando una completamente diferente. 
La idea es mejorarlos, y por mi parte lo más pronto posible, el formato actual hace perder demasiado tiempo en debuggeo, es engorroso, difícil de mantener y viola la mitad de los principios básicos de la programación orientada a objetos; mantenerlo solo porque ofrece una mejora de rendimiento insignificante y por miedo a trabajar de más no me parecen razones válidas.

**Comment by furszy** [comment]
Miedo a trabajar de más..? . estamos en un teléfono Víctor, cualquier optimización es más que válida y más cuando hay libros y documentación que validan lo que estoy diciendo. Lo que estas diciendo de "mejora insignificante" no tiene sustento y más si nos ponemos a pensar que va a ser copiado por todo la plataforma y que esto no va a parar de crecer en base a lo que nosotros hagamos. 
Seguramente ese enum de Plugins no exista más cuando tengamos nuestra plataforma de tokens y podamos validar los plugins sin necesidad de dejarlo en duro, como también todos los enums grandes y engorrosos de Fermat tiendan a desaparecer y a hacerse dinámicos.

**Comment by Yayotron** [comment]
Sí, estoy de acuerdo con que el método del switch es más rápido. Y más aún con que utilizar enums para procesos dinámicos como llevar el control de los plugines/plataformas/etc es darle un mal uso a la  concepción original de este mecanismo. Pero eso de que cualquier optimización es más que válida no lo comparto, si querés irte a los libros podés buscar "Premature optimization is the root of all evil" que es un dicho muy común en este tópico de discusión y hay montones de libros explicándolo. No porque un proceso es más rápido que otro significa que sea mejor, si querés utilizar optimizaciones tan pequeñas a esta escala puedo decirte que el método de los switch consume más memoria y más tiempo del GC por el simple hecho de utilizar Strings hardcodeados para cada case, que como sabemos es un objeto inmutable. Además de que el código compilado de un switch statement utilizando strings llama el método hashCode() que también consume su tiempo. Pero todos esos argumentos a la escala que estamos hablando (en cuanto a cantidad de tiempo y memoria ahorrados) son minúsculos, y en caso de que se quieran tomar en cuenta se debe decidir en un benchmark intenso para comparar uso de memoria/tiempo de GC/tiempo de procesamiento y no yéndonos solo a la teoría porque cada uno tiene sus pros y contras.

**Comment by furszy** [comment]
En eso estoy de acuerdo y comparto (debe ser el café de recién que me puso "bueno"), no todas las optimizaciones son aceptables tempranamente, hasta ahí si queres estamos de acuerdo depende el contexto,
Ese choclo que dijiste lo se y puedo agregar que nuestros enums no deberían usar Strings siempre ,como lo hace la mayoría, para diferenciarse con el tan conocido getByCode() porque ya de por si un enum es un objeto único, aún así sigue siendo más rápido el switch por el simple hecho de que está hecho para eso :p .
Igualmente como dije antes, no es cuestión de optimizar primordial mente esto, es cuestión de que la gente no se acostumbre a hacer las cosas de esta forma por copy y paste cuando debería hacerlo de otra o por lo menos lo mejor que tengamos para entregarles. Si sabemos que este enum de Plugins está mal y va a desaparecer, mejor pongamos le un lindo cartel de "No copiar" y no dejemos que se repitan las cosas así. Nadie está usando el String que viene con cada objeto del el enum por ejemplo, totalmente innecesario.
